lib/events.js

```diff
@@ -84,8 +107,35 @@ EventEmitter.init = function() {
   }
 
   this._maxListeners = this._maxListeners || undefined;
+
+  if (opts && opts.autoCatch) {
+    this[kApply] = applyAndCatch;
+  }
 };
 
+const apply = Reflect.apply;
+function applyAndCatch(handler, that, args) {
+  const promise = apply(handler, that, args);
+  if (promise && typeof promise.then === 'function') {
```

Comment:
Nit, this is not A+ compliant way to handle `then`s since `then` might be a getter that throws on the second attempt. I swear there is a test for this in the Promises/A+ tests.

The "correct" way to do this would be to grab `then` in one line and then use the same reference the second time around.

Also - please handle the case `then` throws here.

Reply:
> Also - please handle the case then throws here.

What do you mean?

Reply:
```js
promise = {
  get then() {
    throw new Error('boom')
  }
}
```

Reply:
and
```js
promise = {
  then() {
    throw new Error('boom')
  }
}
```

Reply:
What @targos said :] Here is a reference from the A+ tests https://github.com/promises-aplus/promises-tests/blob/4786505fcb0cafabc5f5ce087e1df86358de2da6/lib/tests/2.3.3.js#L211

Reply:
I've updated the text. Let me know what you think!